### PR TITLE
Implement loading/saving of Transactionv4

### PR DESCRIPTION
**Pull request by zander**
Allow a transaction of a new version to be loaded using a tagged-format allowing much greater extensibility.

### CODE REVIEW

File: src/consensus/transactionv4.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef TRANSACTION_4_H
+#define TRANSACTION_4_H
+
+namespace Consensus {
+
+enum TxMessageFields {
+    TxEnd = 0,          // BoolTrue
+    TxInPrevHash,       // sha256 (Bytearray)
```
Comment (justusranvier) [inline src/consensus/transactionv4.h]: If you added one more field (TxInPrevHeight) to hold the block height at which the previous transaction was mined, that's the only change that would be necessary to make usable fraud proofs.

Comment (gmaxwell) [reply src/consensus/transactionv4.h]: > that's the only change that would be necessary to make usable fraud proofs

No its not.

Comment (dangquochungne) [inline src/consensus/transactionv4.h]: ok.thanks

File: src/consensus/transactionv4.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    TxOutValue,         // PositiveNumber (in satoshis)
+    TxOutScript,        // bytearray
+    LockByBlock,        // PositiveNumber
+    LockByTime,         // PositiveNumber
+    ScriptVersion       // PositiveNumber
+};
+
+}
+
+#endif
```
Comment (dangquochungne) [inline src/consensus/transactionv4.h]: hj

File: src/consensus/transactionv4.h (hunk 1 above)
Comment (justusranvier) [reply src/consensus/transactionv4.h]: @gmaxwell Did you accidentally post on the wrong thread? It doesn't appear to have any relationship to my comment.

File: src/primitives/transaction.h [hunk 3]
```diff
@@ -195,6 +198,59 @@ class CTxOut
...
+            *const_cast<std::vector<CTxIn>*>(&tx.vin),
+            *const_cast<std::vector<CTxOut>*>(&tx.vout));
+    } else {
+        CSerActionUnserialize ser_action;
+        READWRITE(*const_cast<std::vector<CTxIn>*>(&tx.vin));
+        READWRITE(*const_cast<std::vector<CTxOut>*>(&tx.vout));
+        READWRITE(*const_cast<uint32_t*>(&tx.nLockTime));
+    }
+}
+
```
Comment (greatwolf) [inline src/primitives/transaction.h]: Why are there so many `const_cast` in this function?

File: src/serialize.h [hunk 4]
```diff
@@ -375,6 +378,160 @@ I ReadVarInt(Stream& is)
...
+            return boost::get<uint64_t>(data);
+        return 0;
+    }
+    uint64_t longData() {
+        if (data.which() == 0)
+            return boost::get<int32_t>(data);
+        else if (data.which() == 2)
+            return boost::get<uint64_t>(data);
+        return 0;
+    }
```
Comment (greatwolf) [inline src/serialize.h]: Maybe refactor `intData` and `longData` into one template function? eg. `int intData() { return getDataAs<int>(); }`

File: src/serialize.h [hunk 5]
```diff
@@ -375,6 +378,160 @@ I ReadVarInt(Stream& is)
...
+        switch (format) {
+        case NegativeNumber:
+            data = ReadVarInt<Stream, int32_t>(s) * -1;
+            break;
+        case PositiveNumber: {
+            int64_t tmp = ReadVarInt<Stream, int64_t>(s);
+            if (tmp > 0xFFFFFF)
+                data = (uint64_t) tmp;
+            else
+                data = (int32_t) tmp;
```
Comment (greatwolf) [inline src/serialize.h]: What's the reason for preferring C-style cast over C++? Is it just for brevity? Also why the magic number `0xFF FF FF`? Why not `0x7F FF FF FF` or use `numeric_limits<int32_t>::max()`?

File: src/primitives/transaction.h (hunk 3 above)
Comment (zander) [reply src/primitives/transaction.h]: This is historical; the CTransaction uses public const members for whatever stupid reason. Its also the reason there is a CMutableTransaction.

File: src/serialize.h (hunk 5 above)
Comment (zander) [reply src/serialize.h]: i'll fix

File: src/serialize.h (hunk 4 above)
Comment (zander) [reply src/serialize.h]: I'd rather keep the knowledge of what is stored in the variant local in the class. Your suggestion would throw an exaction if a caller didn't exactly use the 2 types of ints supported here.

**Comment by zander** [comment]
Merged, not sure why it didn't close.

**Pull request closed without merging**